site-src/geps/gep-851.md

```diff
@@ -0,0 +1,102 @@
+# GEP-851: Allow Multiple Certificate Refs per Gateway Listener
+
+* Issue: [#851](https://github.com/kubernetes-sigs/gateway-api/issues/851)
+* Status: Implementable
+
+## TLDR
+
+Replace `CertificateRef` field with a `CertificateRefs` field in Gateway
+Listeners.
+
+## Goals
+
+* Allow certs to be attached for multiple unique Hostnames for the same
+  Listener. This is most helpful in the following situations:
+  * The hostname on the Listener is empty or unspecified allowing Routes to
+    attach arbitrary hostnames to the Listener with Routes.
+  * The hostname on the Listener is a wildcard and Route owners want to attach
+    multi-level subdomains. For example, `*.example.com` on the Listener and
+    `app.dev.example.com` on a Route.
+* Enable specifying RSA and ECDSA certs on the same Listener.
+* Improve cert upgrades by allowing new and old certs to be attached to the same
+  Listener.
+
+## Non-Goals
+
+* Be too prescriptive about how multiple certificates are handled for a Gateway
+  Listener. Each implementation may handle this slightly differently.
+
+## Introduction
+
+When we removed certificate references from HTTPRoute ([GEP
+746](/geps/gep-746/)) we lost the capability to attach multiple certificates to
+a Listener. This GEP proposes adding that capability back, directly to the
+Gateway Listener this time.
+
+In addition to helping with the self service approach, where app owners may
+attach multiple hostnames (and certs) to a single Listener, this can help with
+other use cases:
+
+* Attaching RSA and ECDSA certs.
+* Attaching both new and old certs during an upgrade.
+
+## API
+
+The `CertificateRef` field in `GatewayTLSConfig` would be replaced with the
+following `CertificateRefs` field:
+
+```go
+    // CertificateRefs contains a series of references to Kubernetes objects that
+    // contains TLS certificates and private keys. These certificates are used to
+    // establish a TLS handshake for requests that match the hostname of the
+    // associated listener.
+    //
+    // When multiple TLS certificates are specified, SNI matching MUST be used
+    // to determine the appropriate certificate to use.
+    //
+    // Implementations SHOULD verify that there is an intersection between the
```

Comment:
This assumes the control plane can always introspect the certificate. Is that always the case? I can imagine a lot of cases where this would not be the case.

Istio used to be this way actually, although its not anymore.

Reply:
To clarify does this mean that Istio can no longer derive details from certs?

Reply:
This basically demands that every control plane unwrap certificates referenced here and check the CN and SANs. It *should* be possible for everything to pull the certs from Kuberrnetes themselves (since you need to have read access to them to be able to watch them anyway), unwrap them enough to check the CN and SAN, and then throw the other info away.

For Contour, this is extra functionality we'll need to add, but it is doable.

Reply:
What if we start with a simpler version of this that doesn't require unwrapping? Up to 2 certificates may be attached to a listener, 1 RSA and 1 EC. I don't want to give up on the more complex self-service use cases here, but we may not need to solve them in this first v1a2 release.

Reply:
Istio can implement it today, for kind=secret. Previously though the data plane read the Secret directly.

For other kinds (Especially something like a file reference mounting an HSM or similar) the control plane wouldn't have any clue. Or CSI driver, etc etc.

Overall I think it will come back to bite us if we require control plane can see the cert.

For dual rsa+ecdsa maybe it's fine, although it's not clear to me how common that use case is to introduce this complexity

Reply:
If a control plane can't or won't examine the certificates, this is a SHOULD, so that implementation would sill be conforming.

This SHOULD aligns with [RFC 8446](https://datatracker.ietf.org/doc/html/rfc8446#section-4.4.2.2).

Personally, I think that the less we say in the standard about how this works, the better. There's a lot of divergence across proxy implementations and TLS changes can move very rapidly. It's good to keep the implementation flexibility as long as users have this general guidance about what to put in a secret.

Reply:
If I don't inspect the contents then how would we know which one to present to clients? It seems like we would need to just pick a random one (or make Envoy smarter to do the selection dynamically?)

Reply:
Sorry, what I meant was that some theoretical control plane (not you specifically) might not examine it. Maybe it would make sense for it do dump them all down to the data plane and let that figure it out.

Reply:
Would it be too strict to ask the control plane to unwrap the certificates and check the CN and SAN, what if people just want to use a certificate with different SAN (as a placeholder) on the staging env? Maybe it should be a preference but not a striction.

Reply:
The current wording of "Should" denotes a recommendation and not a requirement (https://datatracker.ietf.org/doc/html/rfc2119#section-3), which I think is what you're suggesting.

Reply:
We may want to turn SHOULD to MAY